README.md
# UCB_W205
Backup of my all projects for the W205-Fundamentals of Data Engineering Course - Spring 2020


project-2-aanikooyan/history.txt
  150  git status
  151  git commit -m 'First commit to the assignment branch - Part 1' 
  152  git push origin assignment
  153  exit
  154  ls
  155  pwd
  156  clear
  157  dkcer -ps
  158  docker ps
  159  docker ps -a
  160  docker ps
  161  docker ps -a
  162  docker rm -f jolly_volted
  163  docker rm -f jolly_volhard
  164  docker ps -a
  165  clear
  166  docker images
  167  clear
  168  exit
  169  docker ps -a
  170  sudo chown -R jupyter:jupyter ~/w205
  171  cd ~/w205/course-content
  172  git pull --all
  173  cd
  174  docker run -it --rm -v ~/w205:/w205 midsw205/base:latest bash
  175  docker run -it -v ~/w205:/w205 midsw205/base:latest bash
  176  docker pull midsw205/base:latest
  177  docker pull midsw205/base:0.1.8
  178  pwd
  179  docker pull midsw205/base:0.1.9
  180  docker pull redis
  181  docker pull confluentinc/cp-zookeeper:latest
  182  docker pull confluentinc/cp-kafka:latest
  183  docker pull midsw205/spark-python:0.0.5
  184  clear
  185  docker pull midsw205/spark-python:0.0.6
  186  docker pull midsw205/cdh-minimal:latest
  187  docker pull midsw205/hadoop:0.0.2
  188  docker pull midsw205/presto:0.0.1
  189  clear
  190  docker ps -a
  191  docker ps
  192  exit
  193  ls
  194  cd ~w205
  195  cdw205
  196  cd w205
  197  ls
  198  cd signup-aanikooyan/
  199  ls
  200  cat README.mdy
  201  ls
  202  cd project-1-aanikooyan/
  203  ls
  204  clear
  205  ls
  206  git branches
  207  git branch
  208  git checkout assignment
  209  ls
  210  rm project1_queries.md
  211  ls
  212  git add -a
  213  git add -A
  214  git status
  215  git commit -m 'resubmission ofPart1'
  216  git push origin assignment
  217  git add README.md
  218  git commit -m 'solving the table printing issue in README'
  219  git push origin assignment
  220  git add README.md
  221  git commit -m 'solving the table printing issue in README'
  222  git push origin assignment
  223  git add README.md
  224  git commit -m 'solving the table printing issue in README'
  225  git push origin assignment
  226  git add README.md
  227  git commit -m 'solving the table printing issue in README'
  228  git push origin assignment
  229  git add README.md
  230  git commit -m 'solving the table printing issue in README'
  231  git push origin assignment
  232  clear
  233  git add README.md
  234  git commit -m 'solving the table printing issue in README'
  235  git push origin assignment
  236  git add README.md
  237  git commit -m 'solving the table printing issue in README'
  238  git push origin assignment
  239  git add README.md
  240  git commit -m 'Some formatting enhancement for Part 1'
  241  git push origin assignment
  242  git add README.md
  243  git commit -m 'Some formatting enhancement for Part 1'
  244  git push origin assignment
  245  exit
  246  clear
  247  ls -a
  248  cd w205
  249  ls
  250  cd signup-aanikooyan/
  251  ls
  252  cd project-1-aanikooyan/
  253  ls
  254  git branch
  255  git checkout assignment
  256  git status
  257  git add README.md
  258  git commit -m 'questions for Part 2 added'
  259  git push origin assignment
  260  bq query --use_legacy_sql=false '
  261          SELECT count(*)
  262          FROM
  263             `bigquery-public-data.san_francisco.bikeshare_trips`'
  264  clear
  265  bq query --use_legacy_sql=false '
  266              SELECT count(distinct trip_id) AS Total_number_of_trips
  267              FROM 
  268                  `bigquery-public-data.san_francisco.bikeshare_trips`'
  269  bq query --use_legacy_sql=false '
  270              SELECT MIN(start_date) AS earliest_start_date, MAX(end_date) AS latest_end_date
  271              FROM 
  272                  `bigquery-public-data.san_francisco.bikeshare_trips`
  273  bq query --use_legacy_sql=false '
  274              SELECT MIN(start_date) AS earliest_start_date, MAX(end_date) AS latest_end_date
  275              FROM 
  276                  `bigquery-public-data.san_francisco.bikeshare_trips`'
  277  bq query --use_legacy_sql=false '
  278              SELECT count(DISTINCT bike_number) AS total_bikes
  279              FROM 
  280                  `bigquery-public-data.san_francisco.bikeshare_trips`'
  281  clear
  282  exit
  283  bq query --use_legacy_sql=false '
  284              SELECT
  285                COUNT(trip_id) AS trips,
  286                CASE
  287                  WHEN EXTRACT(hour 
  288                    FROM start_date) >= 6 
  289                    AND EXTRACT(hour 
  290                    FROM start_date) <= 10 
  291                    THEN 'morning trips'
  292                  WHEN EXTRACT(hour
  293                    FROM start_date) >= 15
  294                    AND EXTRACT(hour
  295                    FROM start_date) <= 19 
  296                    THEN 'afternoon trips'
  297                  ELSE 'other times trips'
  298                END
  299                AS trip_time
  300              FROM
  301                `bigquery-public-data.san_francisco.bikeshare_trips`
  302              GROUP BY
  303                trip_time'
  304  clear
  305  bq query --use_legacy_sql=false '
  306              SELECT
  307                COUNT(trip_id) AS trips,
  308                CASE
  309                  WHEN EXTRACT(hour 
  310                    FROM start_date) >= 6 
  311                    AND EXTRACT(hour 
  312                    FROM start_date) <= 10 
  313                    THEN 'morning trips'
  314                  WHEN EXTRACT(hour
  315                    FROM start_date) >= 15
  316                    AND EXTRACT(hour
  317                    FROM start_date) <= 19 
  318                    THEN 'afternoon trips'
  319                  ELSE 'other times trips'
  320                END
  321                AS trip_time
  322              FROM
  323                `bigquery-public-data.san_francisco.bikeshare_trips`
  324              GROUP BY
  325                trip_time'

project-3-aanikooyan/game_api.py
#!/usr/bin/env python
import json
from kafka import KafkaProducer
from flask import Flask, request

app = Flask(__name__)
producer = KafkaProducer(bootstrap_servers='kafka:29092')


def log_to_kafka(topic, event):
    event.update(request.headers)
    producer.send(topic, json.dumps(event).encode())


@app.route("/")
def default_response():
    default_event = {'event_type': 'default'}
    log_to_kafka('events', default_event)
    return "This is the default response!\n"


@app.route("/purchase_a_sword")
def purchase_a_sword():
    purchase_sword_event = {'event_type': 'purchase_sword'}
    log_to_kafka('events', purchase_sword_event)
    return "Sword Purchased!\n"


@app.route("/join_a_guild")
def join_a_guild():
    join_guild_event = {'event_type': 'join_guild',
                                            'description': 'very nice guild'}
    log_to_kafka('events', join_guild_event)
    return "Guild Joined!\n"

@app.route("/purchase_a_knife")
def purchase_a_knife():
    purchase_knife_event = {'event_type': 'purchase_knife',
                                            'description': 'very sharp knife'}
    log_to_kafka('events', purchase_knife_event)
    return "Knife Purchased!\n"


full-streaming-stack/write_swords_stream.py
#!/usr/bin/env python
"""Extract events from kafka and write them to hdfs
"""
import json
from pyspark.sql import SparkSession
from pyspark.sql.functions import udf, from_json
from pyspark.sql.types import StructType, StructField, StringType


def purchase_sword_event_schema():
    """
    root
    |-- Accept: string (nullable = true)
    |-- Host: string (nullable = true)
    |-- User-Agent: string (nullable = true)
    |-- event_type: string (nullable = true)
    |-- timestamp: string (nullable = true)
    """
    return StructType([
        StructField("Accept", StringType(), True),
        StructField("Host", StringType(), True),
        StructField("User-Agent", StringType(), True),
        StructField("event_type", StringType(), True),
    ])


@udf('boolean')
def is_sword_purchase(event_as_json):
    """udf for filtering events
    """
    event = json.loads(event_as_json)
    if event['event_type'] == 'purchase_sword':
        return True
    return False


def main():
    """main
    """
    spark = SparkSession \
        .builder \
        .appName("ExtractEventsJob") \
        .getOrCreate()

    raw_events = spark \
        .readStream \
        .format("kafka") \
        .option("kafka.bootstrap.servers", "kafka:29092") \
        .option("subscribe", "events") \
        .load()

    sword_purchases = raw_events \
        .filter(is_sword_purchase(raw_events.value.cast('string'))) \
        .select(raw_events.value.cast('string').alias('raw_event'),
                raw_events.timestamp.cast('string'),
                from_json(raw_events.value.cast('string'),
                          purchase_sword_event_schema()).alias('json')) \
        .select('raw_event', 'timestamp', 'json.*')

    sink = sword_purchases \
        .writeStream \
        .format("parquet") \
        .option("checkpointLocation", "/tmp/checkpoints_for_sword_purchases") \
        .option("path", "/tmp/sword_purchases") \
        .trigger(processingTime="15 seconds") \
        .start()

    sink.awaitTermination()


if __name__ == "__main__":
    main()
